lib/AST/RequirementMachine/RequirementMachineRequests.cpp

```diff
@@ -413,6 +413,12 @@ RequirementSignatureRequestRQM::evaluate(Evaluator &evaluator,
       machine->computeRequirementDiagnostics(errors, proto->getLoc());
       diagnoseRequirementErrors(ctx, errors,
                                 AllowConcreteTypePolicy::NestedAssocTypes);
+
+      if (machine->getConcreteType(proto->getSelfInterfaceType(),
```

Comment:
Do you need to do this for each protocol in a connected component? (I know, it's a ridiculous example that nobody will hit) Maybe try something like this

```swift
protocol P1 where Self == Int {
  associatedtype T : P2
}

protocol P2 where Self == Int {
  associatedtype T : P1
}
```

The requirement signature of P1 and P2 is computed simultaneously.

I think the source location used for conflict diagnostics might be suspect too; we seem to always use proto->getLoc() as well.

Reply:
> I think the source location used for conflict diagnostics might be suspect too; we seem to always use proto->getLoc() as well.

I know, right now conflict diagnostics always use the location of the signature itself. Instead, it should try to use the location of the conflicting requirement inside the signature using the structural requirements in the rewrite system.